nomos-services/blend/src/core/tests/mod.rs

```diff
@@ -0,0 +1,181 @@
+        })
+        .await
+        .unwrap();
+
+    // Prepare an epoch handler with the mock chain service that always returns the
+    // same epoch state.
+    let mut epoch_handler = EpochHandler::new(
+        TestChainService,
+        settings.time.epoch_transition_period_in_slots,
+    );
+
+    // Prepare dummy Overwatch resources.
+    let (overwatch_handle, _overwatch_cmd_receiver, state_updater, _state_receiver) =
+        dummy_overwatch_resources();
+
+    // Initialize the service.
+    let (
+        remaining_session_stream,
+        remaining_clock_stream,
+        current_public_info,
+        crypto_processor,
+        blending_token_collector,
+        current_recovery_checkpoint,
+        message_scheduler,
+        backend,
+        rng,
+    ) = initialize::<
+        NodeId,
+        TestBlendBackend,
+        TestNetworkAdapter,
+        TestChainService,
+        MockCoreAndLeaderProofsGenerator,
+        MockProofsVerifier,
+        RuntimeServiceId,
+    >(
+        settings.clone(),
+        membership_stream,
+        clock_stream,
+        &mut epoch_handler,
+        overwatch_handle.clone(),
+        None,
+        state_updater,
+    )
+    .await;
+
+    // Run the event loop of the service in a separate task.
+    let settings_cloned = settings.clone();
+    let join_handle = tokio::spawn(async move {
+        let secret_pol_info_stream =
+            post_initialize::<OncePolStreamProvider, RuntimeServiceId>(&overwatch_handle).await;
+
+        run_event_loop(
+            inbound_relay,
+            blend_message_stream,
+            remaining_clock_stream,
+            secret_pol_info_stream,
+            remaining_session_stream,
+            &settings_cloned,
+            backend,
+            TestNetworkAdapter,
+            epoch_handler,
+            message_scheduler,
+            rng,
+            blending_token_collector,
+            crypto_processor,
+            current_public_info,
+            current_recovery_checkpoint,
+        )
+        .await;
+    });
+
+    // Send a new session with the same membership.
+    membership_info.session_number += 1;
+    membership_sender
+        .send(membership_info.clone())
+        .await
+        .unwrap();
+
+    // Since the node is still core in the new session,
+    // the service must keep running even after a session transition period.
+    sleep(settings.time.session_transition_period().mul_f64(1.5)).await;
+    assert!(!join_handle.is_finished());
+
+    // Send a new session with a new membership smaller than minimal size
+    membership_info.membership = new_membership(minimal_network_size.checked_sub(1).unwrap()).0;
+    membership_info.session_number += 1;
+    membership_sender.send(membership_info).await.unwrap();
+
+    // Since the network is smaller than the minimal size,
+    // the service must stop after a session transition period.
+    timeout(
+        settings.time.session_transition_period().mul_f64(1.5),
```

Comment:
Can we test that any logic in the session transition session is executed, somehow? With this condition we are testing, we only know that after 1.5x the transition period the function exits, but we don't know exactly when. We can use, for instance, `backend.complete_session_transition()` to be notified when the function is called. Or use the blending token collector to make sure it is called before quitting.